src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs

```diff
@@ -1,18 +1,74 @@
-﻿using System.Threading.Tasks;
+﻿using System;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.Identity.Client;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Azure.SignalR
 {
     internal class AccessKey
     {
-        public string Value { get; }
-        public string Id { get; }
+        private readonly TaskCompletionSource<bool> _initializedTcs = new TaskCompletionSource<bool>();
 
-        public Task AuthorizedTask => Task.CompletedTask;
+        public string Value => _accessKey;
 
-        public AccessKey(string key)
+        public string Id => _kid;
+
+        private volatile string _accessKey;
+
+        private volatile string _kid;
+
+        public Task<bool> InitializedTask => _initializedTcs.Task;
+
+        public bool Initialized => InitializedTask.IsCompleted && InitializedTask.Result;
+
+        public AccessKey(string key = null)
+        {
+
+            if (!string.IsNullOrEmpty(key))
+            {
+                _accessKey = key;

```

Comment:
what happens when aad not enabled while key is null?

When key is not null, what if msi is enabled? can it happen?